### PR TITLE
Replace jQuery shorthand event methods

### LINKED ISSUE #5198: Do not use jQuery shorthand event methods in jQuery
Throughout our JavaScript codebase, a lot of code makes use of jQuery methods that according to TypeScript hinting aredeprecated in 3.3.

We should replace all these with `on` or `trigger` as appropriate.

For example

```
$('.selector').click(clickFn);
$('.btn').click();
```
becomes

```
$('.selector').on('click', clickFn);
$('.btn').trigger('click');
```

Methods to fix:
- []  click
- [] mouseover
- [] mouseout
- [] mouseup
- [] submit

### Evidence / Screenshot (if possible)

Screen Shot 2021-05-23 at 9 43 32 AM

### Steps to Reproduce

1. Go to ...
2. Do ...

* Actual: 
* Expected: 

### Details

- **Logged in (Y/N)?** 
- **Browser type/version?** 
- **Operating system?** 
- **Environment (prod/dev/local)?** prod

### Stakeholders
@jdlrobson

### PR DESCRIPTION
**Pull request by Yashs911**
Closes #5198 

Replaced the following shorthand event methods: 

- [X]  click
- [X]  mouseover
- [X]  mouseout
- [X]  mouseup
- [X]  submit

### Technical

Please check the following changes:

- `<a href="javascript:;" onclick="document.forms[\'logout\'].submit()">`
- `<a href="javascript:;" onclick="document.forms[\'logout\'].trigger('submit')">`

- `\$container.find(".work-checkbox").click(update_my_lists);`
- `\$container.find(".work-checkbox").on('click', update_my_lists);`

- `return this.mouseover(handleHover).mouseout(handleHover);`
- `return this.on('mouseover', handleHover).on('mouseout', handleHover);`

### Stakeholders

@jdlrobson

### CODE REVIEW

**Comment by jdlrobson** [comment]
thank you!

**Comment by jdlrobson** [comment]
Will review this a little later in the week. Out of interest are there any eslint rules we could use to detect these usages and prevent further ones being added in future?

**Comment by Yashs911** [comment]
Maybe we can use https://github.com/wikimedia/eslint-plugin-no-jquery

**Comment by jdlrobson** [comment]
Neat that looks promising.

**Comment by Yashs911** [comment]
Rule: https://github.com/wikimedia/eslint-plugin-no-jquery/blob/master/docs/rules/no-event-shorthand.md

**Comment by jdlrobson** [comment]
let's add it!

**Comment by Yashs911** [comment]
I will create a new PR for this or should I add a commit to this one?

**Comment by jdlrobson** [comment]
Adding here would help me review this change but a new PR also works. Up to you!

**Comment by Yashs911** [comment]
I followed the documentation but I guess I missed something but when I run `npm run lint:js` it does not flag any errors. I even changed a few .click events but it didn't flag any error. So if you could take a look...

File: package.json [hunk 1]
```diff
@@ -36,7 +36,8 @@
     "core-js": "3.11.1",
     "css-loader": "5.2.4",
     "datatables.net-dt": "1.10.24",
-    "eslint": "6.8.0",
+    "eslint": "^6.8.0",
+    "eslint-plugin-no-jquery": "^2.6.0",
```
Comment (jdlrobson) [inline package.json:40]: Drop ^ here and line above so that we use a fixed version

File: .eslintrc.json [hunk 2]
```diff
@@ -1,5 +1,11 @@
 {
   "root": true,
+  "plugins": [
+    "eslint-plugin-no-jquery"
+  ],
+  "extends": [
+    "plugin:no-jquery/deprecated"
+  ],
```
Comment (jdlrobson) [inline .eslintrc.json:8]: In rules below you will also need to add:
```
    "no-jquery/no-event-shorthand": "error",
```

**Review by jdlrobson — CHANGES_REQUESTED** [review]
I'm seeing a few more errors when the event is setup:
```
/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/SearchBar.js
   81:9   error  Prefer .on or .trigger to .change  no-jquery/no-event-shorthand
  125:9   error  Prefer .on or .trigger to .resize  no-jquery/no-event-shorthand
  132:17  error  Prefer .on or .trigger to .focus   no-jquery/no-event-shorthand

/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/SearchUtils.js
  146:9  error  Prefer .on or .trigger to .change  no-jquery/no-event-shorthand

/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/add-book.js
  2:5  error  Prefer .on or .trigger to .click  no-jquery/no-event-shorthand

/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/add_new_field.js
  29:13  error  Prefer .on or .trigger to .change  no-jquery/no-event-shorthand
  49:21  error  Prefer .on or .trigger to .focus   no-jquery/no-event-shorthand

/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/edit.js
    9:5   error  Prefer .on or .trigger to .focus  no-jquery/no-event-shorthand
  145:5   error  Prefer .on or .trigger to .keyup  no-jquery/no-event-shorthand
  176:17  error  Prefer .on or .trigger to .focus  no-jquery/no-event-shorthand
  182:17  error  Prefer .on or .trigger to .focus  no-jquery/no-event-shorthand

/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/index.js
  239:45  error  Prefer .on or .trigger to .select  no-jquery/no-event-shorthand

/Users/jrobson/git/openlibrary/openlibrary/plugins/openlibrary/js/ol.js
  39:5  error  Prefer .on or .trigger to .scroll  no-jquery/no-event-shorthand

```

File: .eslintrc.json (hunk 2 above)
Comment (jdlrobson) [reply .eslintrc.json:9]: My bad. I thought this rule was not included in the "plugin:no-jquery/deprecated". Turns out we just had two extends keys :)

CI is failing now with my latest edit to rectify this.

**Comment by Yashs911** [comment]
@jdlrobson how did you run the plugin, because when I try to run `npm run lint:js`. I don't get any errors.

**Comment by jdlrobson** [comment]
Did you pull down 4c2980a ? After `npm install` they should show up....

**Comment by Yashs911** [comment]
> Did you pull down [4c2980a](https://github.com/internetarchive/openlibrary/commit/4c2980acb0ad22bafd71fd41ac5a68c8dded02fb) ? After `npm install` they should show up....

No, I will do that. but to resolve the package-lock.json conflict should I use `npm install --package-lock-only`

**Comment by Yashs911** [comment]
I think I successfully solved the conflict but I doubt I might have messed up so if you can check the package-lock.json file

**Review by jdlrobson — APPROVED** [review]

**Comment by jdlrobson** [comment]
We should fix the warnings in a follow up. Should I open a new ticket?

**Comment by Yashs911** [comment]
> We should fix the warnings in a follow up. Should I open a new ticket?

I fixed the one related to jquery-event-shorthand and jquery-ready-shorthand in this PR https://github.com/internetarchive/openlibrary/pull/5201/commits/2327bc94c897aefea369ea7388c370b6a9ca8793 but I was to sure if we want to fix other errors

File: openlibrary/templates/account/create.html [hunk 3]
```diff
@@ -33,7 +33,7 @@ <h1>$_("Sign Up")</h1>
 $if ctx.user:
     $def user_link(): <a href="$ctx.user.key">$ctx.user.displayname</a>
     <p>$:_("You are already logged into Open Library as %(user)s.", user=str(user_link()))</p>
-    <p>$:_('If you\'d like to create a new, different Open Library account, you\'ll need to <a href="javascript:;" onclick="document.forms[\'logout\'].submit()">log out</a> and start the signup process afresh.')</p>
+    <p>$:_('If you\'d like to create a new, different Open Library account, you\'ll need to <a href="javascript:;" onclick="document.forms[\'logout\'].trigger('submit')">log out</a> and start the signup process afresh.')</p>
```
Comment (jamesachamp) [inline openlibrary/templates/account/create.html:36]:
```suggestion
    <p>$:_('If you\'d like to create a new, different Open Library account, you\'ll need to <a href="javascript:;" onclick="document.forms[\'logout\'].trigger(\'submit\')">log out</a> and start the signup process afresh.')</p>
```

Single quotes surrounding `submit` must be escaped, or the `create.html` page will throw an error.

File: openlibrary/templates/login.html [hunk 4]
```diff
@@ -12,7 +12,7 @@ <h1>$_("Log In")</h1>
 
   $if ctx.user:
       <p>$:_("You are already logged into Open Library as %(user)s.", user='<a href="%s">%s</a>' % (ctx.user.key, ctx.user.displayname))</p>
-      <p>$:_('If you\'d like to create a new, different Open Library account, you\'ll need to <a href="javascript:;" onclick="document.forms[\'logout\'].submit()">log out</a> and start the signup process afresh.')</p>
+      <p>$:_('If you\'d like to create a new, different Open Library account, you\'ll need to <a href="javascript:;" onclick="document.forms[\'logout\'].trigger('submit')">log out</a> and start the signup process afresh.')</p>
```
Comment (jamesachamp) [inline openlibrary/templates/login.html:15]:
```suggestion
      <p>$:_('If you\'d like to create a new, different Open Library account, you\'ll need to <a href="javascript:;" onclick="document.forms[\'logout\'].trigger(\'submit\')">log out</a> and start the signup process afresh.')</p>
```

If single quotes surrounding `submit` are not escaped, `login.html` throws an error.

File: openlibrary/templates/lib/nav_head.html [hunk 5]
```diff
@@ -129,7 +129,7 @@
           <li><a href="$homepath()/account">$_("Settings")</a></li>
           <li>
             <form name="logout" action="/account/logout" method="post">
-              <a href="#" onclick="document.forms['logout'].submit()">$_("Log out")</a>
+              <a href="#" onclick="document.forms['logout'].trigger('submit')">$_("Log out")</a>
```
Comment (jamesachamp) [inline openlibrary/templates/lib/nav_head.html:132]:
```suggestion
              <a href="#" onclick="document.forms['logout'].submit()">$_("Log out")</a>
```

`document.forms.logout` is not a jQuery object, and does not have the function `trigger`.  This is causing log outs to fail.  Suggest changing back to original code, which used the vanilla `form.submit()` method.

File: openlibrary/templates/login.html (hunk 4 above)
Comment (jamesachamp) [reply openlibrary/templates/login.html:15]: Disregard previous suggestion.  This should be rolled back to the original code, as `document.forms.logout` is not a jQuery object.

File: openlibrary/templates/account/create.html (hunk 3 above)
Comment (jamesachamp) [reply openlibrary/templates/account/create.html:36]: Disregard previous suggestion.  This should be rolled back to the original code, as `document.forms.logout` is not a jQuery object.

**Comment by jamesachamp** [comment]
I noticed that this was causing login, log out, and account creation issues on my local instance.  Sorry to request changes after this has been merged, but it's vital that the suggested changes are made before this is deployed to production.

**Comment by jamesachamp** [comment]
Marked everything as resolved, as I am about to open a PR to address the account issues.

File: openlibrary/templates/account/create.html (hunk 3 above)
Comment (jdlrobson) [reply openlibrary/templates/account/create.html:36]: dispatchEvent is the native equivalent

**Pull request merged**